File: Ant/chapter2/cpp/zero_one_knapsack_problem_dp.cpp
#include <algorithm>
#include <fstream>
#include <stdio.h>

#define MAX_N 100
#define MAX_W 10'000

using namespace std;

int N, W;
int w[MAX_N]{}, v[MAX_N]{};
int dp[MAX_N + 1][MAX_W + 1]{};

int main() {
  ifstream ifs("../testset/zero_one_knapsack_problem/test1.txt");
  ifs >> N >> W;
  for (int i = 0; i < N; ++i) {
    ifs >> w[i] >> v[i];
  }
  for (int i = N - 1; i >= 0; --i) {
    for (int j = 0; j <= W; ++j) {
      if (j < w[i]) {
        dp[i][j] = dp[i + 1][j];
      } else {
        dp[i][j] = max(dp[i + 1][j], dp[i + 1][j - w[i]] + v[i]);
      }
    }
  }
  printf("%d\n", dp[0][W]);
}

File: Ant/chapter2/python/zero_one_knapsack_problem_dp.py
def ns(f):
    return next(f).strip()


with open("../testset/zero_one_knapsack_problem/test1.txt", 'r') as f:
    N, W = map(int, ns(f).split())
    P = [list(map(int, ns(f).split())) for _ in range(N)]   # (w, v)

dp = [[0] * (W + 1) for _ in range(N + 1)]
for i in range(N - 1, -1, -1):
    for j in range(W + 1):
        if j < P[i][0]:
            dp[i][j] = dp[i + 1][j]
        else:
            dp[i][j] = max(dp[i + 1][j], dp[i + 1][j - P[i][0]] + P[i][1])
print(dp[0][W])
